p2p/service/gossip_listener_test.go

```diff
@@ -112,8 +118,54 @@ func Test_AddListener_notSynced(t *testing.T) {
 	assert.NoError(t, n1.Broadcast(context.TODO(), "channel1", []byte{}))
 	assert.NoError(t, n1.Broadcast(context.TODO(), "channel2", []byte{}))
 
-	assert.Equal(t, atomic.LoadInt32(&channelCount), int32(0))
-	assert.Equal(t, atomic.LoadInt32(&secondChannel), int32(0))
+	assert.Equal(t, int32(0), atomic.LoadInt32(&channelCount))
+	assert.Equal(t, int32(0), atomic.LoadInt32(&secondChannel))
+}
+
+func TestListenerConcurrency(t *testing.T) {
+	net := NewSimulator()
+	n1 := net.NewNode()
+	var listenCount int32
+	listenFn := func() bool {
+		atomic.AddInt32(&listenCount, 1)
+		return true
+	}
+	conf := config.DefaultTestConfig()
+	l := NewListener(n1, &syncMock{true, listenFn}, func() bool { return true }, conf, log.NewDefault(n1.Info.ID.String()))
+	defer l.Stop()
+
+	var channelCount int32
 
-	l.Stop()
+	releaseChan := make(chan struct{})
+	handlerFn := func(ctx context.Context, data GossipMessage, syncer Fetcher) {
+		<-releaseChan
+		atomic.AddInt32(&channelCount, 1)
+	}
+
+	l.AddListener(context.TODO(), "channel1", priorityq.Mid, handlerFn)
+
+	assert.Equal(t, 2, conf.MaxGossipRoutines)
+
+	// broadcast several messages. expect the first two to call a handler, then the rest should block.
+	assert.NoError(t, n1.Broadcast(context.TODO(), "channel1", []byte{}))
+	assert.NoError(t, n1.Broadcast(context.TODO(), "channel1", []byte{}))
+	assert.NoError(t, n1.Broadcast(context.TODO(), "channel1", []byte{}))
+	assert.NoError(t, n1.Broadcast(context.TODO(), "channel1", []byte{}))
+
+	checkVal := func(expectedVal int32) func() bool {
+		return func() bool {
+			return atomic.LoadInt32(&listenCount) == expectedVal
+		}
+	}
+	assert.Eventually(t, checkVal(2), time.Second, 10*time.Millisecond)
+
+	// release one handler
+	releaseChan <- struct{}{}
+	assert.Eventually(t, checkVal(3), time.Second, 10*time.Millisecond)
+	releaseChan <- struct{}{}
+	assert.Eventually(t, checkVal(4), time.Second, 10*time.Millisecond)
+	releaseChan <- struct{}{}
```

Comment:
curious what the significance of adding data to releaseChan manually and check that listenCount doesn't increase?

Reply:
listenCount is a running count of how many listeners ran. we queue up several and given a concurrency factor of 2, we initially expect this to be set to 2. then we release them one by one and expect this to tick up to 4 and stay there. does that make sense?